docs/docker.md

```diff
@@ -15,24 +16,74 @@ You want a dockerfile that is structured something similar to this
 ```Dockerfile
 # Dockerfile
 
-# Build container
-FROM continuumio/miniconda:latest as conda
+# -----------------
+# Builder container
+# -----------------
+FROM continuumio/miniconda:latest as builder
 
 ADD conda-linux-64.lock /locks/conda-linux-64.lock
 RUN conda create -p /opt/env --copy --file /locks/conda-linux-64.lock
 
+# -----------------
 # Primary container
+# -----------------
 FROM gcr.io/distroless/base-debian10
 # copy over the generated environment
-COPY --from=conda /opt/env /opt/env
+COPY --from=builder /opt/env /opt/env
+ENV PATH="/opt/env/bin:${PATH}"
 ```
 
-To get this to work nicely generate the platform specific lock run
+To get this to work nicely generate the platform specific lock run something
+like this in your shell
 
-```shell
+```bash
+# Update the lockfile
 conda lock --format explicit --platform linux-64
-docker build .
+# build the image
+docker build -t myimagename:mytag .
 ```
 
 This will ensure that your conda dependencies used in this docker container are
 always exactly reproducible.
+
+## conda-lock inside a build container
+
+You can also use conda-lock with a build-container style system if you mnake use of
+the `--copy` flag from `conda-lock install`
+
+```Dockerfile
+# -----------------
+# Builder container
+# -----------------
+FROM condaforge/mambaforge:4.14.0-0 as builder
+
+COPY environment.yml /docker/environment.yml
+
+RUN . /opt/conda/etc/profile.d/conda.sh && \
+    mamba create --name lock && \
+    conda activate lock && \
+    mamba env list && \
+    mamba install --yes pip conda-lock>=1.2.2 setuptools wheel && \
+    conda-lock lock \
```

Comment:
If I'm following correctly, then `conda-lock.yml` only lives in the Docker cache where it can't be tracked with version control. Do we really want to recommend this?

Reply:
hence the two approaches in this

Reply:
I feel like I'm missing something obvious here... I think I don't know what is a "build-container style system".

Two reasons I can think of off the top of my head for locking:
1. Control explicitly when dependencies are updated
2. Having an explicit record of the versions installed

It seems like neither of these applies in this case. (For 2. at least there's a file in the container, but if you have to go into the container to see it, you might as well run `conda env export`.)

What's the benefit to this approach over installing directly from `environment.yml`?

Reply:
So the main benefit is more when you are using something like `earthly` and have a single earthfile specify multiple docker outputs.  

1. This example does NOT deal with this case
2. The explicit record is at least kept in the build-container style build described here, which means you can introspect a bit more easily as to what should be in the env.